### PR TITLE
fix: improve required errors for validation levels

### PR DESCRIPTION
**Pull request by tjallingt**
Remove the `requiredVerified` and `requiredPublish` options from the JSON Schema, the checks are moved to after AJV has checked the schema. This allows us to improve the error message to mention that the error is _because_ of the validation level.

### CODE REVIEW

File: lib/App/index.js [hunk 1]
```diff
@@ -143,6 +120,16 @@ class App {
...
     }
 
+    if (level >= VALIDATION_LEVELS.VERIFIED) {
+      if (appJson.platforms === undefined) {
+        throw new Error('The property `platforms` is required in order to publish a verified app.');
+      }
+
+      if (appJson.support === undefined) {
+        throw new Error('The property `support` is required in order to publish a verified app.');
+      }
```
Comment (RobinBol) [inline lib/App/index.js:130]: This doesn't verify that `platforms` is an array and `support` a valid support string, not sure if that used to be the case though.

File: lib/App/index.js [hunk 2]
```diff
@@ -152,8 +139,10 @@ class App {
           throw new Error(`Forbidden permission: ${permission}`);
         }
 
-        if (permission === 'homey:manager:api' && levelPublish) {
-          console.warn('Warning: using the homey:manager:api permission will require a more thorough review. It may take longer than usual to review your app.');
+        if (permission === 'homey:manager:api') {
+          if (level >= VALIDATION_LEVELS.PUBLISH) {
+            console.warn('Warning: using the homey:manager:api permission will require a more thorough review. It may take longer than usual to review your app.');
```
Comment (RobinBol) [inline lib/App/index.js:144]: Maybe this should use the `Log` method?

File: lib/App/index.js [hunk 3]
```diff
@@ -277,10 +268,20 @@ class App {
...
           await this._validateImages(driver.images, 'driver');
         }
 
+        if (level >= VALIDATION_LEVELS.VERIFIED) {
+          if (driver.platforms === undefined) {
+            throw new Error(`drivers.${driver.id}: property \`platforms\` is required in order to publish a verified app.`);
+          }
+
+          if (driver.connectivity === undefined) {
+            throw new Error(`drivers.${driver.id}: property \`connectivity\` is required in order to publish a verified app.`);
```
Comment (RobinBol) [inline lib/App/index.js:281]: This doesn't verify the value of these properties

File: lib/App/index.js [hunk 4]
```diff
@@ -580,36 +578,43 @@ class App {
...
+    const filteredCardArgs = card.args.filter(argument => {
+      return argument !== firstDeviceArgument;
     });
 
     if (filteredCardArgs.length === 0) return;
 
+    if (level >= VALIDATION_LEVELS.VERIFIED) {
+      for (const argument of filteredCardArgs) {
+        if (argument.title === undefined) {
+          throw new Error(`${errorPath}: property \`title\` for is required for arguments in order to publish a verified app.`);
```
Comment (RobinBol) [inline lib/App/index.js:599]:
```suggestion
          throw new Error(`${errorPath}: property \`title\` is required for arguments in order to publish a verified app.`);
```

File: lib/App/index.js [hunk 5]
```diff
@@ -580,36 +578,43 @@ class App {
...
-      } else {
-        console.warn(`Warning: ${errorPath}.titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.`);
+    if (level >= VALIDATION_LEVELS.VERIFIED) {
+      if (card.titleFormatted === undefined) {
+        throw new Error(`${errorPath}: property \`titleFormatted\` is required in order to publish a verified app.`);
       }
     }
 
+    if (typeof card.titleFormatted === 'undefined') {
+      console.warn(`Warning: ${errorPath}.titleFormatted is missing. Specifying a Flow card's formatted title will be required in the future.`);
```
Comment (RobinBol) [inline lib/App/index.js:615]: `Log` perhaps?

File: package.json [hunk 6]
```diff
@@ -28,7 +28,7 @@
   "config": {
     "npmPublishTagProduction": "latest",
     "npmPublishTagTesting": "beta",
-    "nodeVersion": "8"
+    "nodeVersion": "12"
   },
```
Comment (RobinBol) [inline package.json:32]: I think this config object isn't actually used anymore

File: lib/App/index.js (hunk 1 above)
Comment (tjallingt) [reply lib/App/index.js:130]: The schema takes care of that, the schema just doesn't make sure the property actually exists.

File: lib/App/index.js (hunk 2 above)
Comment (tjallingt) [reply lib/App/index.js:144]: Maybe, don't know if `node-homey-lib` as such a method. Code was already like this so seems fine to keep it like this.

File: lib/App/index.js (hunk 3 above)
Comment (tjallingt) [reply lib/App/index.js:281]: The schema takes care of that, the schema just doesn't make sure the property actually exists.

File: package.json (hunk 6 above)
Comment (tjallingt) [reply package.json:32]: Yea don't think so either, I can remove and add the `"engines"` field if we want that

File: lib/App/index.js (hunk 2 above)
Comment (RobinBol) [reply lib/App/index.js:144]: It does, might be more noticible using colors, but not an issue really: https://github.com/athombv/node-homey/search?q=Log

**Review by RobinBol — APPROVED** [review]

File: helpers/index.js [hunk 7]
```diff
@@ -1,28 +1,18 @@
-/* eslint-disable global-require */
-
 'use strict';
 
-try {
```
Comment (WeeJeWel) [inline helpers/index.js:5]: Why has this been changed? It's for browser & react environments.

File: lib/Signal/index.js [hunk 8]
```diff
@@ -20,26 +20,13 @@ class Signal {
     this._check = this._check.bind(this);
   }
 
-  debug(...args) {
```
Comment (WeeJeWel) [inline lib/Signal/index.js:23]: Why has this been removed?

File: lib/App/index.js [hunk 9]
```diff
@@ -3,35 +3,42 @@
 
 'use strict';
 
-const querystring = require('querystring');
+const {
+  join,
+  extname,
+  basename,
+  dirname,
+} = require('path');
```
Comment (WeeJeWel) [inline lib/App/index.js:11]: Please use `path.join()` in code, otherwise it's hard to trace where a function comes from.

File: lib/App/index.js [hunk 10]
```diff
@@ -3,35 +3,42 @@
...
-  dirname,
+  readBytesFromFile,
 } = require('../../helpers');
 
-const VALIDATION_LEVELS = [
-  'debug',
-  'publish',
-  'verified',
-];
+const VALIDATION_LEVELS = {
```
Comment (WeeJeWel) [inline lib/App/index.js:36]: It's not per se implied that levels follow up on each other. It's better to manually specify them for every check, leaving room for future levels with different requirements.

File: lib/App/index.js [hunk 11]
```diff
@@ -3,35 +3,42 @@
...
 const Signal = require('../Signal');
 const Energy = require('../Energy');
 
+const statAsync = util.promisify(fs.stat);
+const readFileAsync = util.promisify(fs.readFile);
+const readDirAsync = util.promisify(fs.readdir);
+const lstatAsync = util.promisify(fs.lstat);
+const imageSizeAsync = util.promisify(imageSize);
+
 const {
```
Comment (WeeJeWel) [inline lib/App/index.js:32]: Ideally we make a class `Util` with static method `readBytesFromFile`. It's more consistent with our other repositories.

**Review by WeeJeWel — CHANGES_REQUESTED** [review]

**Comment by WeeJeWel** [comment]
@tjallingt I think now is a great moment to create tests that validates a spectrum of apps, where some are on purpose invalid.

**Comment by tjallingt** [comment]
Cleaned up the changes to solve the comments, reverted the change to treat level as a number.
I think adding tests is a good idea, I'll make another PR for that.

**Comment by WeeJeWel** [comment]
👍🏻  I'll wait with merging this until that PR is ready.

**Comment by tjallingt** [comment]
The tests are in https://github.com/athombv/node-homey-lib/pull/330
I rebased this PR on that branch so it is easier to review the changes to the tests (just error message wording)

File: lib/App/index.js [hunk 12]
```diff
@@ -583,33 +595,40 @@ class App {
...
-      return false;
+    const filteredCardArgs = card.args.filter(argument => {
+      return argument !== firstDeviceArgument;
     });
 
     if (filteredCardArgs.length === 0) return;
 
+    if (levelVerified) {
+      for (const argument of filteredCardArgs) {
+        if (argument.title === undefined) {
```
Comment (jeroenwienk) [inline lib/App/index.js:612]: Why are these checks for `undefined` and not also `null`? Bit odd that someone would set a property to `null` but still possible.

**Review by jeroenwienk — APPROVED** [review]

File: lib/App/index.js (hunk 12 above)
Comment (tjallingt) [reply lib/App/index.js:612]: AJV will error if any of these are set to `null` , but `undefined` is allowed when a property is not required.

**Pull request merged**